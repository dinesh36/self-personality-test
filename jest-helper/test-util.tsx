import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import React from 'react';
import { render } from '@testing-library/react';
import { getStore } from '../src/store/store/store';

export const getComponentWithReduxWrapper = (cmp: React.ReactNode) => {
  const store = getStore();
  return renderer.create(
    <Provider store={store}>
      {cmp}
    </Provider>,
  );
};

export const getComponentWithReduxWrapperForReactTestingLib = (cmp: React.ReactNode) => {
  const store = getStore();
  return render(
    <Provider store={store}>
      {cmp}
    </Provider>,
  );
};
