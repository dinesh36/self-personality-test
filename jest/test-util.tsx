import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import React from 'react';
import { getStore } from '../src/store/store/store';

export const getComponentWithReduxWrapper = (cmp: React.ReactNode) => {
  const store = getStore();
  return renderer.create(
    <Provider store={store}>
      {cmp}
    </Provider>,
  );
};
