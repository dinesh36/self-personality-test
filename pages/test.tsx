import type { NextPage } from 'next';

const Test: NextPage = () => {
    return (
        <div>
            test page
        </div>
    )
}

export default Test
