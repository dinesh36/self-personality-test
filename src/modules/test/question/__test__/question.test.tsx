import { getComponentWithReduxWrapper, getComponentWithReduxWrapperForReactTestingLib } from '@jestHelper/test-util';
import { fireEvent } from '@testing-library/react';
import * as redux from 'react-redux';
import Question from '../question';
import { IQuestion } from '../../shared/test.types';
import TEST_ACTIONS from '../../../../store/store/test/test.action-types';

const mockDispatch = jest.fn();

const state = {
  testState: {
    answers: {},
    currentQuestionIndex: 0,
    questions: [1],
  },
};
jest
  .spyOn(redux, 'useSelector')
  .mockImplementation((callback) => callback(state));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

describe('Question Component', () => {
  const mockQuestionId = 123;
  const questionMock: IQuestion = {
    title: 'mock question',
    id: mockQuestionId,
    answerOptions: [
      {
        title: 'answer 1',
        id: 1,
        alphabet: 'B',
      },
      {
        title: 'answer 2',
        id: 2,
        alphabet: 'A',
      },
    ],
  };
  it('Should render the question component', () => {
    const component = getComponentWithReduxWrapper(<Question
      question={questionMock}
      submitTest={() => {}}
    />);
    const cmp = component.toJSON();

    expect(cmp).toMatchSnapshot();
  });

  it('Should get the next question when clicking the next button', () => {
    const mockSubmitFn = jest.fn();
    const cmp = getComponentWithReduxWrapperForReactTestingLib(<Question
      question={questionMock}
      submitTest={mockSubmitFn}
    />);
    const buttonSelector = new RegExp('Finish Test', 'i');
    expect(cmp.queryByText(buttonSelector)).toBeTruthy();
    fireEvent.click(cmp.getByText(buttonSelector));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: TEST_ACTIONS.UPDATE_ANSWER,
      questionId: mockQuestionId,
      answerOptionId: 1,
    });
    expect(mockSubmitFn).toHaveBeenCalled();
  });

  it('Should get the next question when clicking the next button', () => {
    state.testState.questions = [1, 2];
    const cmp = getComponentWithReduxWrapperForReactTestingLib(<Question
      question={questionMock}
      submitTest={() => {}}
    />);
    const buttonSelector = new RegExp('Next', 'i');
    expect(cmp.queryByText(buttonSelector)).toBeTruthy();
    fireEvent.click(cmp.getByText(buttonSelector));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: TEST_ACTIONS.UPDATE_ANSWER,
      questionId: mockQuestionId,
      answerOptionId: 1,
    });
    expect(mockDispatch).toHaveBeenCalledWith({
      type: TEST_ACTIONS.UPDATE_CURRENT_QUESTION,
      newQuestionIndex: 1,
    });
  });

  it('Should get the previous question when clicking the previous button', () => {
    state.testState.questions = [1, 2];
    state.testState.currentQuestionIndex = 1;
    const cmp = getComponentWithReduxWrapperForReactTestingLib(<Question
      question={questionMock}
      submitTest={() => {}}
    />);
    const buttonSelector = new RegExp('Previous', 'i');
    expect(cmp.queryByText(buttonSelector)).toBeTruthy();
    fireEvent.click(cmp.getByText(buttonSelector));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: TEST_ACTIONS.UPDATE_ANSWER,
      questionId: mockQuestionId,
      answerOptionId: 1,
    });
    expect(mockDispatch).toHaveBeenCalledWith({
      type: TEST_ACTIONS.UPDATE_CURRENT_QUESTION,
      newQuestionIndex: 0,
    });
  });
});
