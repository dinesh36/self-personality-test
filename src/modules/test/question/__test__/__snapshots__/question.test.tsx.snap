// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Component Should render the question component 1`] = `
<div
  className="card question"
>
  <div
    className="questionTitle"
  >
    mock question
  </div>
  <div
    className="answerOptions"
  >
    <div
      className="answerOption selected"
      onClick={[Function]}
    >
      <span>
        B
        .  
        answer 1
      </span>
      <img
        alt="selected icon"
        src="images/test/selected-icon.svg"
      />
    </div>
    <div
      className="answerOption "
      onClick={[Function]}
    >
      <span>
        A
        .  
        answer 2
      </span>
    </div>
  </div>
  <div
    className="buttonContainer firstQuestion"
  >
    <button
      className="ant-btn css-dev-only-do-not-override-1n7nwfa ant-btn-primary button undefined"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span>
        Finish Test
      </span>
    </button>
  </div>
</div>
`;
